akka-stream-tests/src/test/scala/akka/stream/scaladsl/CancellablePublisherSinkSpec.scala

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream.testkit.{ TestPublisher, TestSubscriber }
+import akka.stream.ActorMaterializer
+
+import akka.testkit.AkkaSpec
+import akka.stream.testkit.Utils._
+import scala.concurrent.duration._
+
+class CancellablePublisherSinkSpec extends AkkaSpec {
+
+  implicit val materializer = ActorMaterializer()
+
+  "A CancellablePublisherSink" must {
+
+    "work in happy case" in assertAllStagesStopped {
+      val pub = Source.single(1).runWith(Sink.asCancellablePublisher[Int]())
+      val sub = TestSubscriber.probe[Int]
+      pub.subscribe(sub)
+      sub.request(1)
+        .expectNext(1)
+        .expectComplete()
+    }
+
+    "drain buffer when no subscribers attached" in assertAllStagesStopped {
+      val sub1 = TestSubscriber.probe[Int]
+      val sub2 = TestSubscriber.probe[Int]
+      val src = TestPublisher.probe[Int]()
+      val pub = Source.fromPublisher(src).runWith(Sink.asCancellablePublisher[Int]())
+
+      pub.subscribe(sub1)
+      sub1.request(1)
+      sub1.expectNoMsg(200.millis)
+      src.sendNext(1)
+      sub1.expectNext(1).cancel()
+      src.sendNext(2)
+      src.sendNext(3)
+
```

Comment:
there is a race here between enqueueing the sentNext events and enqueueing the subscribe below. I don't see any nice way to avoid this without a sleep though. Any ideas?